File: poly_poly_bot/polymarket/src/trade-source.ts
// Source abstraction for trade detection — data-api, onchain, or hybrid mode

export interface TradeSource {
  name: string;
  start(): Promise<void>;
  stop(): void;
}

import { DataApiSource } from "./data-api-source";
import { OnchainSource } from "./onchain-source";

/** Create trade sources based on TRADE_MONITOR_MODE config. */
export function createSources(mode: string): TradeSource[] {
  switch (mode) {
    case "onchain":
      return [new OnchainSource()];
    case "hybrid":
      return [new OnchainSource(), new DataApiSource()];
    default:
      return [new DataApiSource()];
  }
}

File: poly_poly_bot/src/copy_trading/trade_source.py
"""Trade source factory -- data-api, onchain, or hybrid mode."""

from __future__ import annotations
from typing import Protocol


class TradeSource(Protocol):
    """Protocol for pluggable trade detection sources."""

    name: str

    async def start(self) -> None: ...
    def stop(self) -> None: ...


def create_sources(mode: str) -> list[TradeSource]:
    """Create trade source(s) based on the configured mode.

    Args:
        mode: One of 'data-api', 'onchain', or 'hybrid'.

    Returns:
        List of trade sources to run concurrently.
    """
    from src.copy_trading.data_api_source import DataApiSource
    from src.copy_trading.onchain_source import OnchainSource

    if mode == "onchain":
        return [OnchainSource()]
    elif mode == "hybrid":
        return [OnchainSource(), DataApiSource()]
    else:
        return [DataApiSource()]
